torch_geometric/nn/conv/heat_pyg.py

```diff
@@ -0,0 +1,187 @@
+        src_node_n_edge_feat = torch.cat((emb_edge_feat, src_node_emb), dim=1) # e^+_{ij}
+
+        ## ------------------------ v Eq. 9 v ------------------------ ## 
+        tar_src_node_n_edge_feat = torch.cat((tar_node_emb, src_node_n_edge_feat), dim=1) # eq.9 [ \arrow{h}_{\kappai} | e^+_{i,j}]
+        scores = self.leaky_relu(self.attention_nn(tar_src_node_n_edge_feat))
+        scores_matrix = torch.sparse_coo_tensor(edge_index, scores, torch.Size([node_feats.shape[0], node_feats.shape[0], self.heads])).to_dense().to(node_feats.device)
+        scores_matrix = torch.where(scores_matrix==0.0, torch.ones_like(scores_matrix) * -10000, scores_matrix)
+        attention_matrix = self.soft_max(scores_matrix) # [num_nodes, num_nbrs, heads]]
+        ## =========================================================== ## 
+
+        ## ------------------------ v Eq. 10 v ------------------------ ## 
+        src_n_edge_attr = self.leaky_relu(self.node_update_emb(torch.cat((src_node_emb, edge_attr_emb), dim=1))) # eq.10 W_h[ concat. ]
+        src_n_edge_attr_matrix = torch.sparse_coo_tensor(edge_index, src_n_edge_attr, 
+                                    torch.Size([node_feats.shape[0], node_feats.shape[0], self.out_channels])).to_dense().to(node_feats.device)
+        
+        attention_matrix = attention_matrix.unsqueeze(3).repeat(1,1,1,self.out_channels)
+        src_n_edge_attr_matrix = src_n_edge_attr_matrix.unsqueeze(dim=2).repeat(1, 1, self.heads, 1)
+
+        next_node_feat = torch.sum(torch.mul(attention_matrix,  src_n_edge_attr_matrix), dim=1)
+        ## =========================================================== ## 
+        if self.concat:
+            next_node_feat = torch.flatten(next_node_feat, start_dim=1)
+        else:
+            next_node_feat = torch.mean(next_node_feat, dim=1)
+        return next_node_feat
+
+if __name__ == '__main__':
+    from torch_geometric.data import Data, DataLoader
+
+    ## Pre-define the node_type_set and edge_type_set to cover all the node and edge types in the dataset.
+    node_type_set = ('n1', 'n2', 'n3')
+    edge_type_set = ('e1', 'e2', 'e3', 'e4')
+
+    ## Prepare a data
+    node_features = torch.randn((3, 4))
+    node_type_set = ('n1', 'n2', )
+    edge_type_set = ('e1', 'e2', 'e3')
+
+    node_type_list = ['n1', 'n2', 'n2']
+    edge_type_list = ['e1', 'e3', 'e2', 'e2', 'e3']
+    edge_type = torch.randn((5, 2))
+    edge_attr = torch.randn((5, 2))
+    edge_index = torch.tensor([[0, 0, 1, 2, 0],
+                               [1, 2, 0, 0, 0]])
+    data_1 = Data(  x=node_features, edge_index=edge_index, 
+                    node_type_list=node_type_list, edge_type_list=edge_type_list,
+                    edge_type=edge_type, edge_attr=edge_attr )
+                    
+    node_features = torch.randn((3, 4))
+    
+
+    node_type_list = ['n1', 'n3', 'n2']
+    edge_type_list = ['e1', 'e3', 'e4', 'e2', 'e3']
+    edge_type = torch.randn((5, 2))
```

Comment:
Why is `edge_type` a feature matrix?

Reply:
edge_type does not need to be a feature matrix with floats. I changed edge_type to `edge_type = torch.tensor([[0, 2, 1, 1, 2]], dtype=torch.long)` something like this, indicating 5 edges falls into 3 types (0,1,2).